# Project: Wrangling and Analyze Data

### IMPORT PACKAGES

In [1]:
import pandas as pd
import numpy as np
import requests
import os
import json
import seaborn as sns
import matplotlib.pyplot as plt

### Data Gathering
> 1. twitter-archive-enhanced.csv
> 2. image-predictions.tsv
> 3. tweet-json.txt

In [2]:
# reading twitter archive dataset
df_twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')

2. Use the Requests library to download the tweet image prediction (image_predictions.tsv)

In [3]:
# Reading file from url
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
r = requests.get(url)
with open(url.split('/')[-1], mode = 'wb') as f:
        f.write(r.content)


In [4]:
#Reading extracted file
img_pred_df = pd.read_csv ('image-predictions.tsv', sep='\t')

3. Use the Tweepy library to query additional data via the Twitter API (tweet_json.txt)

In [5]:
# Reading tweet
tweet_list = []
with open('tweet-json.txt') as f:
    for line in f:        
        twit = json.loads(line)
        tweet_id = twit['id']
        retweets =twit['retweet_count']
        likes = twit['favorite_count']
        tweet_list.append({'tweet_id':tweet_id,'retweet_count':retweets,'favorite_count':likes})
df_tweets = pd.DataFrame(tweet_list)


## Assessing Data


#### Visual Assessment

In [6]:
# archive visual assesement

df_twitter_archive.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
249   845306882940190720                    NaN                  NaN   
1212  715342466308784130                    NaN                  NaN   
1479  693582294167244802           6.935722e+17         1.198989e+09   
751   779056095788752897                    NaN                  NaN   
1456  695314793360662529                    NaN                  NaN   

                      timestamp  \
249   2017-03-24 16:10:40 +0000   
1212  2016-03-31 00:58:29 +0000   
1479  2016-01-30 23:51:19 +0000   
751   2016-09-22 20:33:42 +0000   
1456  2016-02-04 18:35:39 +0000   

                                                 source  \
249   <a href="http://twitter.com/download/iphone" r...   
1212  <a href="http://twitter.com/download/iphone" r...   
1479  <a href="http://twitter.com/download/iphone" r...   
751   <a href="http://twitter.com/download/iphone" r...   
1456  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
249   This is Pickles. She's a silly pupper. Thinks ...                  NaN   
1212  This is Oscar. He's a world renowned snowball ...                  NaN   
1479  Personally I'd give him an 11/10. Not sure why...                  NaN   
751   Everybody drop what you're doing and look at t...                  NaN   
1456  This is Colin. He really likes green beans. It...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
249                        NaN                        NaN   
1212                       NaN                        NaN   
1479                       NaN                        NaN   
751                        NaN                        NaN   
1456                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
249   https://twitter.com/dog_rates/status/845306882...                12   
1212  https://twitter.com/dog_rates/status/715342466...                10   
1479                                                NaN                11   
751   https://twitter.com/dog_rates/status/779056095...                13   
1456  https://twitter.com/dog_rates/status/695314793...                10   

      rating_denominator     name doggo floofer  pupper puppo  
249                   10  Pickles  None    None  pupper  None  
1212                  10    Oscar  None    None    None  None  
1479                  10     None  None    None    None  None  
751                   10     None  None    None    None  None  
1456                  10    Colin  None    None    None  None

In [7]:
# image visual assessment

img_pred_df.sample(5)

tweet_id                                            jpg_url  \
841   694905863685980160    https://pbs.twimg.com/media/CaTNMUgUYAAB6vs.jpg   
282   671134062904504320    https://pbs.twimg.com/media/CVBY3e7XIAAAE4Y.jpg   
1430  773191612633579521    https://pbs.twimg.com/media/CrrtqjdXEAINleR.jpg   
682   683834909291606017  https://pbs.twimg.com/ext_tw_video_thumb/68383...   
177   669214165781868544    https://pbs.twimg.com/media/CUmGu7-UcAA0r3O.jpg   

      img_num                p1   p1_conf  p1_dog                p2   p2_conf  \
841         1           bow_tie  0.449268   False          fur_coat  0.139099   
282         1          Shih-Tzu  0.180380    True  golden_retriever  0.180194   
1430        1  Blenheim_spaniel  0.427766    True          Shih-Tzu  0.219256   
682         1       Maltese_dog  0.738449    True        toy_poodle  0.102992   
177         1           minivan  0.435396   False        police_van  0.310143   

      p2_dog                      p3   p3_conf  p3_dog  
841    False     black-footed_ferret  0.082232   False  
282     True      Labrador_retriever  0.173656    True  
1430    True  Welsh_springer_spaniel  0.144614    True  
682     True                 Samoyed  0.023247    True  
177    False                 minibus  0.068201   False

In [8]:
# tweets visual assessment

df_tweets.sample(5)

tweet_id  retweet_count  favorite_count
1781  677673981332312066           1677            3603
1407  699060279947165696           2059            4173
1497  692187005137076224            929            2768
1911  674372068062928900            346             931
1981  672964561327235073            701            1530

### Programatic Assessment

#### PROGRAMMATIC ASSESSMENT FOR TWITTER ARCHIVE

In [9]:
df_twitter_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [10]:
# checking for correct dtype inputs

df_twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

#### PROGRAMMATIC ASSESSMENT FOR IMAGE PREDICTION

In [11]:
img_pred_df.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [12]:
# checking for correct image entries

img_pred_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [13]:
# unique values in image

img_pred_df.nunique()

tweet_id    2075
jpg_url     2009
img_num        4
p1           378
p1_conf     2006
p1_dog         2
p2           405
p2_conf     2004
p2_dog         2
p3           408
p3_conf     2006
p3_dog         2
dtype: int64

In [14]:
# get total count for each image number

img_pred_df.img_num.value_counts()

1    1780
2     198
3      66
4      31
Name: img_num, dtype: int64

### PROGRAMMATIC ASSESSMENT FOR JSON TWEEETS

In [15]:
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2354 non-null   int64
 1   retweet_count   2354 non-null   int64
 2   favorite_count  2354 non-null   int64
dtypes: int64(3)
memory usage: 55.3 KB


In [16]:
df_tweets.describe()

tweet_id  retweet_count  favorite_count
count  2.354000e+03    2354.000000     2354.000000
mean   7.426978e+17    3164.797366     8080.968564
std    6.852812e+16    5284.770364    11814.771334
min    6.660209e+17       0.000000        0.000000
25%    6.783975e+17     624.500000     1415.000000
50%    7.194596e+17    1473.500000     3603.500000
75%    7.993058e+17    3652.000000    10122.250000
max    8.924206e+17   79515.000000   132810.000000

In [17]:
all_columns = pd.Series(list(df_twitter_archive) + list(img_pred_df) + list(df_tweets))
all_columns[all_columns.duplicated()]

17    tweet_id
29    tweet_id
dtype: object

### Quality issues



> 1. archive - missing variable (expanded_urls) incomplete data entry of 2297 instead of 2356

> 2. archive - replace "a" and "o" in name column to None in name column

> 3. archive - 'timestamp' contains two info of both 'date' and 'time', hence it will be splitted

> 4. archive - 'date' and timestamp on 'archive' should be converted to Datetime dtype

> 5. archive - 'source' should be extracted from the hyperlink

> 6. archive - all retweet and reply entries are inconsistent and will be dropped 

> 7. image - drop the jpg_url, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog columns

> 8. image - dropping rows in df_image_predictions with p2_dog == false


### Tidiness issues


> 1. archive - merge 'doggo, floofer, pupper and puppo' into 'dogbreed' columns

> 2. both 'json' and 'image' will be merged into 'archive' as a single dataframe for the project


## Cleaning Data
In this section, we would be cleaning all the documented quality and tidiness issues.

In [18]:
# Make copies of original pieces of data

df_twitter_archive_clean = df_twitter_archive.copy()
img_pred_df_clean = img_pred_df.copy()
df_tweets_clean = df_tweets.copy()

### Quality Issue:

#### Define: 

> **ISSUE 1:** archive_clean - cleaning all retweeted rows with null values

#### Code 

In [19]:
#identify all list with retweet values and drop 

retweeted_tweets = df_twitter_archive_clean[df_twitter_archive_clean.retweeted_status_id.notnull()].index.tolist()

df_twitter_archive_clean.drop(retweeted_tweets, inplace = True)

#### Test

In [20]:
df_twitter_archive_clean[df_twitter_archive_clean.retweeted_status_id.notnull()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

#### Define: 

> **ISSUE 2:** archive_clean - updating 'expanded_urls' missing data

In [21]:
df_twitter_archive_clean['expanded_urls']=df_twitter_archive_clean['expanded_urls'].fillna(0)

#### Test

In [22]:
len(df_twitter_archive_clean['expanded_urls'])

2175

#### Define:

> **ISSUE 2:** updating name entries 'a' and 'o' to None

#### Code 

In [23]:
df_twitter_archive_clean.loc[df_twitter_archive_clean.name == 'a', 'name'] = 'None'

df_twitter_archive_clean.loc[df_twitter_archive_clean.name == 'o', 'name'] = 'None'

#### Test

In [24]:
df_twitter_archive_clean.name

0        Phineas
1          Tilly
2         Archie
3          Darla
4       Franklin
          ...   
2351        None
2352        None
2353        None
2354        None
2355        None
Name: name, Length: 2175, dtype: object

#### Define

> **ISSUE 3:** seperating the 'timestamp' to 'date' and 'time' on tweet_id



#### Code

In [26]:
df_twitter_archive_clean['timestamp']=pd.date_range('2017-08-01 16:23:56',periods = 2175)
df_twitter_archive_clean['date']=pd.to_datetime(df_twitter_archive_clean['timestamp']).dt.date
df_twitter_archive_clean['time']=pd.to_datetime(df_twitter_archive_clean['timestamp']).dt.time
df_twitter_archive_clean = df_twitter_archive_clean.drop('timestamp', axis=1)

#### Test

In [27]:
df_twitter_archive_clean.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'source',
       'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo',
       'date', 'time'],
      dtype='object')

#### Define

> **ISSUE 4:** converting date and time to datetime dtype


#### Code 

In [28]:
df_twitter_archive_clean['date'] = pd.to_datetime(df_twitter_archive_clean['date'])


#### Test 

In [29]:
df_twitter_archive_clean.dtypes

tweet_id                               int64
in_reply_to_status_id                float64
in_reply_to_user_id                  float64
source                                object
text                                  object
retweeted_status_id                  float64
retweeted_status_user_id             float64
retweeted_status_timestamp            object
expanded_urls                         object
rating_numerator                       int64
rating_denominator                     int64
name                                  object
doggo                                 object
floofer                               object
pupper                                object
puppo                                 object
date                          datetime64[ns]
time                                  object
dtype: object

#### Define 

> **ISSUE 5:** Extracting source from hyperlink

#### Code

In [30]:
df_twitter_archive_clean['source'] =df_twitter_archive_clean['source'].str.replace(r'<[^<>]*>', '', regex = True)

#### Test

In [31]:
df_twitter_archive_clean.source.sample(4)

870     Twitter for iPhone
1610    Twitter for iPhone
1334    Twitter for iPhone
100     Twitter for iPhone
Name: source, dtype: object

#### Define

> ** ISSUE 6:** Dropping all retweets and reply in archive

#### Code

In [32]:
df_twitter_archive_clean.drop(['in_reply_to_status_id','in_reply_to_user_id','retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp', 'expanded_urls'], axis = 1, inplace = True)

#### Test

In [33]:
df_twitter_archive_clean.head()

tweet_id              source  \
0  892420643555336193  Twitter for iPhone   
1  892177421306343426  Twitter for iPhone   
2  891815181378084864  Twitter for iPhone   
3  891689557279858688  Twitter for iPhone   
4  891327558926688256  Twitter for iPhone   

                                                text  rating_numerator  \
0  This is Phineas. He's a mystical boy. Only eve...                13   
1  This is Tilly. She's just checking pup on you....                13   
2  This is Archie. He is a rare Norwegian Pouncin...                12   
3  This is Darla. She commenced a snooze mid meal...                13   
4  This is Franklin. He would like you to stop ca...                12   

   rating_denominator      name doggo floofer pupper puppo       date  \
0                  10   Phineas  None    None   None  None 2017-08-01   
1                  10     Tilly  None    None   None  None 2017-08-02   
2                  10    Archie  None    None   None  None 2017-08-03   
3                  10     Darla  None    None   None  None 2017-08-04   
4                  10  Franklin  None    None   None  None 2017-08-05   

       time  
0  16:23:56  
1  16:23:56  
2  16:23:56  
3  16:23:56  
4  16:23:56

#### Define

> ** ISSUE 7:** based on the info from image.describe p1 details are more relevant hence dropping other info

#### Code

In [34]:
img_pred_df_clean.drop(["jpg_url", "p2", "p2_conf", "p2_dog", "p3", "p3_conf", "p3_dog"], axis = 1, inplace = True)

#### Test

In [35]:
img_pred_df_clean.head()

tweet_id  img_num                      p1   p1_conf  p1_dog
0  666020888022790149        1  Welsh_springer_spaniel  0.465074    True
1  666029285002620928        1                 redbone  0.506826    True
2  666033412701032449        1         German_shepherd  0.596461    True
3  666044226329800704        1     Rhodesian_ridgeback  0.408143    True
4  666049248165822465        1      miniature_pinscher  0.560311    True

#### Define

> ** ISSUE 8:** Dropping all rows == false on p2_dog 

#### Code

In [36]:
img_pred_df_clean.index[img_pred_df_clean["p1_dog"] == False]


Int64Index([   6,    8,   17,   18,   21,   22,   25,   29,   33,   43,
            ...
            1992, 2012, 2013, 2021, 2022, 2026, 2046, 2052, 2071, 2074],
           dtype='int64', length=543)

In [37]:
img_pred_df_clean.drop(img_pred_df_clean[img_pred_df_clean["p1_dog"] == False].index, inplace = True)

#### Test

In [38]:
img_pred_df_clean.index[img_pred_df_clean["p1_dog"] == False]

Int64Index([], dtype='int64')

### Tidiness

#### Define

> ** ISSUE 1:** We are going to concate the 'json' table to the 'archive' table


#### Code

In [39]:
# columns to be affected in the archive table 
Dogbreed = df_twitter_archive_clean[['tweet_id','name','doggo','floofer','pupper','puppo']]

# replace all non with nan and joining all data set into dog_breed

df_dogbreed = (Dogbreed.set_index(['tweet_id','name'])
              .replace('None',np.nan)
              .apply(lambda x: ','.join(x.dropna()), axis=1)
              .reset_index(name='dog_breed'))

# merging dog_breed on tweet_id and name
df_twitter_archive_clean =df_twitter_archive_clean.merge(df_dogbreed, on=['tweet_id','name'],how='inner')

# dropping the existing data names
df_twitter_archive_clean.drop(['doggo','floofer','pupper','puppo'],axis=1, inplace=True)

#### Test

In [40]:
df_twitter_archive_clean.columns


Index(['tweet_id', 'source', 'text', 'rating_numerator', 'rating_denominator',
       'name', 'date', 'time', 'dog_breed'],
      dtype='object')

#### Define

> ** ISSUE 2:** concate both 'archive_clean' and 'image_clean' into archive as a single dataframe 

#### Code

In [41]:
# merging jason on archive

df_twitter_archive_clean = pd.merge(df_twitter_archive_clean, df_tweets_clean[['tweet_id','retweet_count','favorite_count']], on=['tweet_id'],how='inner')

# merging image on archive

df_twitter_archive_clean = pd.merge(df_twitter_archive_clean, img_pred_df_clean[['tweet_id','img_num','p1','p1_conf','p1_dog']], on=['tweet_id'],how='inner')


#### Test

In [42]:
df_twitter_archive_clean.sample(5)

tweet_id              source  \
236   826598365270007810  Twitter for iPhone   
1416  667455448082227200  Twitter Web Client   
809   706310011488698368  Twitter for iPhone   
994   686034024800862208  Twitter for iPhone   
1102  678396796259975168  Twitter for iPhone   

                                                   text  rating_numerator  \
236   This is Pawnd... James Pawnd. He's suave af. 1...                13   
1416  This is Reese and Twips. Reese protects Twips....                 7   
809   Here's a very sleepy pupper. Thinks it's an ai...                12   
994   Your fav crew is back and this time they're em...                12   
1102  These little fellas have opposite facial expre...                12   

      rating_denominator   name       date      time dog_breed  retweet_count  \
236                   10  Pawnd 2018-06-24  16:23:56                     2709   
1416                  10  Reese 2023-04-25  16:23:56                       66   
809                   10   None 2020-09-16  16:23:56    pupper           9034   
994                   10   None 2021-06-21  16:23:56                     1324   
1102                  10   None 2021-12-06  16:23:56                      478   

      favorite_count  img_num               p1   p1_conf  p1_dog  
236            11117        1   French_bulldog  0.628119    True  
1416             203        1  Tibetan_terrier  0.676376    True  
809            23443        1         Pembroke  0.698165    True  
994             3424        1       Great_Dane  0.236920    True  
1102            1731        2         Pembroke  0.956180    True

## Storing Data
Save gathered, assessed, and cleaned master dataset to a CSV file named "twitter_archive_master.csv".

In [43]:
df_twitter_archive_clean.to_csv('twitter_archive_master.csv',index=False)

## Analyzing and Visualizing Data
In this section, analyze and visualize your wrangled data. You must produce at least **three (3) insights and one (1) visualization.**

In [ ]:
df_twitter_archive_master=pd.read_csv('twitter_archive_master.csv')

In [ ]:
df_twitter_archive_master.head()

In [ ]:
df_twitter_archive_master.describe()

### Insights:

> 1. Dog breed image have the highest count for the study period?

> 2. Highest tweet source used in the platform.

> 3. most predicted dog breed for the period of investigation?



In [ ]:
# dog image with the highest count 

df_twitter_archive_master.p1.value_counts()

In [ ]:
# source with most used plaftorm

df_twitter_archive_master.source.value_counts()

In [ ]:
# dog breed most recorded

df_twitter_archive_master.dog_breed.value_counts()

### Visualization

In [ ]:
# visualising of dog image with the highest value count

viz=df_twitter_archive_master

plt.figure(figsize=(22,10))
sns.countplot(x='p1', order= viz.p1.value_counts().index[0:-1], data=viz);
plt.title('Dog predictions in tweet archive',fontsize=30)
plt.xlabel('Dog Breed Predictions',fontsize=20)
plt.xticks(rotation=90)
plt.ylabel('Prediction Value Counts',fontsize=20);

> From the image above it will be obeserved that golden retreiver has the highest value counts at 150 followed by labrador retriever at 100 and other dog breeds counts as low as 1